src/coreclr/tools/Common/TypeSystem/Ecma/SymbolReader/UnmanagedPdbSymbolReader.cs

```diff
@@ -30,20 +33,87 @@ public static PdbSymbolReader TryOpenSymbolReaderForMetadataFile(string metadata
     /// </summary>
     public sealed class UnmanagedPdbSymbolReader : PdbSymbolReader
     {
+#if NET6_0_OR_GREATER
+        private static int CLRCreateInstance(ref Guid clsid, ref Guid riid, out ICLRMetaHost ppInterface)
+        {
+            int hr = CLRCreateInstance(ref clsid, ref riid, out IntPtr ptr);
+            ppInterface = hr == 0
+                ? (ICLRMetaHost)ClrMetaHostWrapperCache.Instance.GetOrCreateObjectForComInstance(ptr, CreateObjectFlags.UniqueInstance)
+                : null;
+            return hr;
+
+            [DllImport("mscoree.dll")]
+            static extern int CLRCreateInstance(ref Guid clsid, ref Guid riid, out IntPtr ptr);
+        }
+#else
         [DllImport("mscoree.dll")]
         private static extern int CLRCreateInstance([In] ref Guid clsid, [In] ref Guid riid, [Out, MarshalAs(UnmanagedType.Interface)] out ICLRMetaHost ppInterface);
+#endif
 
+#if !NET6_0_OR_GREATER
         [Guid("d332db9e-b9b3-4125-8207-a14884f53216")]
         [InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
         [ComVisible(true)]
+#endif
         interface ICLRMetaHost
         {
+#if NET6_0_OR_GREATER
+            public static readonly Guid IID = new Guid("d332db9e-b9b3-4125-8207-a14884f53216");
+#endif
+
             [PreserveSig]
             int GetRuntime([In, MarshalAs(UnmanagedType.LPWStr)] String pwzVersion, [In] ref Guid riid, [Out, MarshalAs(UnmanagedType.Interface)] out ICLRRuntimeInfo ppRuntime);
 
             // Don't need any other methods.
         }
 
+#if NET6_0_OR_GREATER
+        private sealed class ClrMetaHostWrapperCache : ComWrappers
```

Comment:
I think it would be better to have a single `ComWrappers` implementation rather than one per each interface.

Reply:
The problem I ran into there was it meant I had to linearly search for the appropriate wrapper via QueryInterface, as opposed to just statically creating the right wrapper. It also risks me making a subtle mistake in the ComWrapper GetOrCreateObject implementation, where the static invocation probably wouldn't.

Is there a recommended way to avoid this? Whole-assembly ComWrapper didn't make sense to me outside of dynamic implementation.

Reply:
ping @AaronRobinsonMSFT

Reply:
> Is there a recommended way to avoid this? Whole-assembly ComWrapper didn't make sense to me outside of dynamic implementation.

The recommendation here is to source generate so mistakes like the above is minimal. I'm not sure how impacting the QI expense is in this case. If you find this clearer then let's keep it.